federationclient.go

```diff
@@ -124,7 +144,10 @@ func (ac *FederationClient) SendJoin(
 		var v1Res []json.RawMessage
 		err = ac.doRequest(ctx, v1req, &v1Res)
 		if err == nil && len(v1Res) == 2 {
-			err = json.Unmarshal(v1Res[1], &res)
+			if err = json.Unmarshal(v1Res[1], &res); err != nil {
```

Comment:
Shouldn't this be unmarshalling to `intermediate`? It's kinda gross that `process()` takes args by side-effects as opposed to just passing them in which may have caught this? Can we modify `process` to accept `intermediate` as an arg rather than doing this in closures perhaps?

Reply:
The main reason I didn't do this was because it avoided having to give a named type to `intermediate`, but yes, you're right that it is unmarshalling to the wrong place there for the fallback. I'll fix it.

Reply:
You don't _have_ to give named types, `func(a struct{Foo string}) { ... }` works. Kinda gross though.